### PR TITLE
Configure Qdrant API-key authentication (1.2.2)

### PR DESCRIPTION
**Pull request by leynos**
## Summary

Implements roadmap task `(1.2.2)` for Qdrant API-key authentication.
The branch now provisions a first-boot raw key at `/etc/repovec/qdrant-api-key`, restricts it to `repovec:repovec` mode `0400`, refreshes the rootful Podman secret `repovec-qdrant-api-key`, and injects that secret into Qdrant as `QDRANT__SERVICE__API_KEY`.

Roadmap task: `(1.2.2)`
ExecPlan: [docs/execplans/1-2-2-configure-qdrant-api-key-authentication.md](https://github.com/leynos/repovec-appliance/blob/1-2-2-configure-qdrant-api-key-authentication/docs/execplans/1-2-2-configure-qdrant-api-key-authentication.md)

## Review walkthrough

- Start with the ExecPlan for the final implementation record, decisions, tolerance approval, and validation evidence.
- Review `packaging/libexec/repovec-qdrant-api-key`, `packaging/systemd/repovec-qdrant-api-key.service`, `packaging/sysusers.d/repovec.conf`, and `packaging/systemd/qdrant.container` for the host-side authentication wiring.
- Review `crates/repovec-core/src/appliance/qdrant_quadlet/` and `crates/repovec-core/tests/` for the static validator, rstest coverage, and rstest-bdd scenarios.
- Review `docs/users-guide.md`, `docs/developers-guide.md`, `docs/repovec-appliance-technical-design.md`, and `docs/roadmap.md` for the operator and maintainer documentation updates.

## Validation

- `make fmt`: passed.
- `make markdownlint`: passed.
- `make nixie`: passed.
- `make check-fmt`: passed.
- `make lint`: passed, including the Whitaker sub-check.
- `make test`: passed, with 71 nextest tests and 24 doctests.

## Notes

- `shellcheck` is not installed in this environment, so the helper was validated through static Rust asset tests and repository gates instead.
- The live systemd/Podman Qdrant smoke test was not run in this coding environment; the ExecPlan includes exact reviewer commands for a disposable appliance-style host.

## Summary by Sourcery

Configure Qdrant to use API-key authentication via a provisioned secret and extend validation and documentation accordingly.

New Features:
- Introduce a first-boot provisioning service and helper that generate a Qdrant API key, store it under /etc/repovec/qdrant-api-key with restricted permissions, and expose it as a Podman secret for the Qdrant container.
- Enforce a static contract that Qdrant receives its API key via the QDRANT__SERVICE__API_KEY environment variable sourced from the repovec-qdrant-api-key Podman secret.

Enhancements:
- Extend the Qdrant Quadlet validator with checks for API-key provisioning dependencies, required Podman secret wiring, and disallowed inline API-key environment variables.
- Add static tests that assert the provisioning systemd unit, helper script, and sysusers configuration match the expected API-key contract.
- Refine developer documentation around CI/docs gating and describe the extended appliance validation and testing patterns.

Documentation:
- Update the users guide and technical design documents to describe Qdrant API-key provisioning, access patterns, and authenticated/unauthenticated behaviour.
- Mark roadmap item 1.2.2 as complete and add an ExecPlan document detailing the API-key authentication design and validation steps.
- Refresh documentation index entries to reference the new Qdrant API-key ExecPlan and improve formatting for existing sections.

Tests:
- Add rstest unit tests and rstest-bdd scenarios that cover the Qdrant API-key Quadlet contract, including required secrets, dependencies, and rejection of inline API keys.

### CODE REVIEW

File: crates/repovec-core/src/appliance/qdrant_quadlet/mod.rs [hunk 1]
```diff
@@ -220,3 +243,79 @@ fn validate_auto_update(parsed: &ParsedQuadlet) -> Result<(), QdrantQuadletError
...
+    }
+
+    has_env_type && has_target
+}
+
+fn validate_no_inline_api_key_environment(
+    parsed: &ParsedQuadlet,
+) -> Result<(), QdrantQuadletError> {
+    for environment in parsed.values(CONTAINER_SECTION, "Environment") {
+        if environment.starts_with(&format!("{QDRANT_API_KEY_ENVIRONMENT_VARIABLE}=")) {
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/repovec-core/src/appliance/qdrant_quadlet/mod.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: packaging/libexec/repovec-qdrant-api-key [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        --shell /usr/sbin/nologin \
+        "${REPOVEC_USER}"
+}
+
+generate_key_file() {
+    tmp_file="$(mktemp "${CONFIG_DIR}/qdrant-api-key.XXXXXX")"
+    trap 'rm -f "${tmp_file}"' EXIT HUP INT TERM
+
+    od -An -N32 -tx1 /dev/urandom | tr -d ' \n' >"${tmp_file}"
+    printf '\n' >>"${tmp_file}"
```
[the remark this answers is not in the document]
Comment (leynos) [reply packaging/libexec/repovec-qdrant-api-key]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (3 errors, 5 warnings, 1 inconclusive)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Snapshot test circumvents validator. Errors constructed directly bypass normalisation tested. Provisioning tests rely on static string assertions without functional verification. | Drive snapshots through real validate\_qdrant\_quadlet failures with mutated Quadlet contents. Enhance provisioning tests with functional assertions beyond string checks. |
| Module-Level Documentation | ❌ Error | Module docstrings exist but lack required relationship documentation. All modules explain purpose/function, but most don't document their relationships to other components as required. | Add relationship documentation to module docstrings: mod.rs should reference submodules; test modules should reference each other; parser.rs should explain validation integration; error.rs should document usage context. |
| Security And Privacy | ❌ Error | Error enum derives Debug, exposing plaintext secrets in assertion diffs. Tests use assert\_eq! on error variants with plaintext assignments. Podman secret creation lacks error handling. | Redact InlineApiKeyEnvironmentDisallowed's environment field from Debug. Drive snapshots through validator. Add error handling for podman secret create failure. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Provisioning tests only assert file strings; never execute helper. Snapshots construct errors directly, not via validator. No integration tests. | Execute helper in tests. Test key generation edge cases. Refactor snapshots to use mutated Quadlets. Add integration test for user, key, and secret creation. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | API key snapshot tests use inline (@) snapshots and bypass validator path. Should drive Display messages through validate\_qdrant\_quadlet() and use committed external snapshots per developers-guide. | Refactor api\_key\_tests.rs: consolidate validation failures into parametrised #\[rstest\] harness. Commit Display output to .snap files following existing tests.rs pattern. |
| Domain Architecture | ⚠️ Warning | PR violates domain architecture: (1) Error snapshots bypass validator path, (2) Infrastructure failures masked with exit 0, (3) Secrets leak via derived Debug on error variant. | Redact environment to variable name. Drive snapshots through validate\_qdrant\_quadlet. Fail closed on Podman errors unless "in use" confirmed. |
| Observability | ⚠️ Warning | Missing error handling: `podman secret create` can fail silently whilst systemd reports success. No alerts, metrics, or operational context for diagnosing failures. Secret may leak in Debug output. | Add error checking for `podman secret create`. Log specific failure reasons. Verify secrets redacted in Debug output. Add systemd health checks or metrics. |
| Concurrency And State | ⚠️ Warning | Key file and Podman secret lack mutual exclusion. Concurrent provisioning races on generation and secret lifecycle. Silent exit 0 on removal failure allows stale credential. No concurrency tests. | Add file locking for key file. Protect Podman secret operations. Fail on secret removal failure. Test concurrent invocations and partial failure scenarios. |
| Unit Architecture | ❓ Inconclusive | No result was produced after verification. Marking as INCONCLUSIVE. | Re-run the check or adjust instructions to produce a final result. |

File: crates/repovec-core/src/appliance/qdrant_quadlet/error.rs [hunk 3]
```diff
@@ -71,6 +71,30 @@ pub enum QdrantQuadletError {
...
+    /// The API-key Podman secret is present but wired to the wrong secret or target.
+    IncorrectApiKeySecret {
+        /// The unexpected secret value.
+        secret: String,
+    },
+    /// The Quadlet exposes the Qdrant API key through an inline environment entry.
+    InlineApiKeyEnvironmentDisallowed {
+        /// The offending environment value.
+        environment: String,
+    },
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/repovec-core/src/appliance/qdrant_quadlet/error.rs:102]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai please create new GitHub issues for the following:

### Out-of-scope items — proposed GitHub issues

The following items from the warning checks are out of scope for this PR. Shall I open GitHub issues for them?

| Item | Scope |
| --- | --- |
| **Concurrency And State** — file locking for key file and Podman secret; fail on removal failure; concurrent invocation tests | Separate hardening PR |
| **Testing (Unit)** — integration test for system-user, key-file, and Podman secret creation lifecycle | Requires a container/VM test harness |
| **Observability** — systemd health checks or metrics endpoint for provisioning failures | Future roadmap item |
